server/src-lib/Hasura/GraphQL/Schema/Table.hs

```diff
@@ -0,0 +1,172 @@
+-- permissions for.
+tableSelectColumnsEnum
+  :: (MonadSchema n m, MonadRole r m, MonadTableInfo r m)
+  => QualifiedTable
+  -> SelPermInfo
+  -> m (Maybe (Parser 'Both n PGCol))
+tableSelectColumnsEnum table selectPermissions = do
+  tableName <- qualifiedObjectToName table
+  columns   <- tableSelectColumns table selectPermissions
+  let enumName    = tableName <> $$(G.litName "_select_column")
+      description = Just $ G.Description $
+        "select columns of table " <>> table
+  pure $ P.enum enumName description <$> nonEmpty
+    [ ( define $ pgiName column
+      , pgiColumn column
+      )
+    | column <- columns
+    ]
+  where
+    define name =
+      P.mkDefinition name (Just $ G.Description "column name") P.EnumValueInfo
+
+-- | Table update columns enum
+--
+-- Parser for an enum type that matches the columns of the given
+-- table. Used for conflict resolution in "insert" mutations, among
+-- others. Maps to the table_update_column object.
+--
+-- Return Nothing if there's no column the current user has "update"
+-- permissions for.
+tableUpdateColumnsEnum
+  :: (MonadSchema n m, MonadRole r m, MonadTableInfo r m)
+  => QualifiedTable
+  -> UpdPermInfo
+  -> m (Maybe (Parser 'Both n PGCol))
+tableUpdateColumnsEnum table updatePermissions = do
+  tableName <- qualifiedObjectToName table
+  columns   <- tableUpdateColumns table updatePermissions
+  let enumName    = tableName <> $$(G.litName "_update_column")
+      description = Just $ G.Description $
+        "update columns of table " <>> table
+  pure $ P.enum enumName description <$> nonEmpty
+    [ ( define $ pgiName column
+      , pgiColumn column
+      )
+    | column <- columns
+    ]
+  where
+    define name =
+      P.mkDefinition name (Just $ G.Description "column name") P.EnumValueInfo
+
+tablePermissions
+  :: forall m n r. (MonadSchema n m, MonadTableInfo r m, MonadRole r m)
+  => QualifiedTable
+  -> m (Maybe RolePermInfo)
+tablePermissions table = do
+  roleName  <- askRoleName
+  tableInfo <- askTableInfo table
+  pure $ getRolePermInfo roleName tableInfo
+
+tableSelectPermissions
+  :: forall m n r. (MonadSchema n m, MonadTableInfo r m, MonadRole r m)
+  => QualifiedTable
+  -> m (Maybe SelPermInfo)
+tableSelectPermissions table = (_permSel =<<) <$> tablePermissions table
+
+tableUpdatePermissions
+  :: forall m n r. (MonadSchema n m, MonadTableInfo r m, MonadRole r m)
+  => QualifiedTable
+  -> m (Maybe UpdPermInfo)
+tableUpdatePermissions table = (_permUpd =<<) <$> tablePermissions table
+
+tableDeletePermissions
+  :: forall m n r. (MonadSchema n m, MonadTableInfo r m, MonadRole r m)
+  => QualifiedTable
+  -> m (Maybe DelPermInfo)
+tableDeletePermissions table = (_permDel =<<) <$> tablePermissions table
+
+tableSelectFields
+  :: forall m n r. (MonadSchema n m, MonadTableInfo r m, MonadRole r m)
+  => QualifiedTable
+  -> SelPermInfo
+  -> m [FieldInfo]
+tableSelectFields table permissions = do
+  -- TODO(PDV): memoize this?
```

Comment:
What's the status of this TODO? I suppose we don't need to memoize anything to avoid infinite loops (based on the fact that the test suite runs fine). Ditto for `tableUpdateColumns`.

Reply:
I was more concerned about performance; those get called a lot during schema generation. But I don't think that warrants a TODO (in hindsight I doubt they actually are a bottleneck), I'll remove those comments.